go/ephemeral/device_ek_storage.go

```diff
@@ -189,3 +212,55 @@ func (s *DeviceEKStorage) MaxGeneration(ctx context.Context) (maxGeneration keyb
 	}
 	return maxGeneration, nil
 }
+
+func (s *DeviceEKStorage) DeleteExpired(ctx context.Context) (expired []keybase1.EkGeneration, err error) {
+	defer s.G().CTrace(ctx, "DeviceEKStorage#DeleteExpired", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+
+	cache, err := s.getCache(ctx)
+	if err != nil {
+		return expired, err
+	}
+
+	keyMap := make(keyExpiryMap)
+	for generation, deviceEK := range cache {
+		keyMap[generation] = deviceEK.Metadata.Ctime
+	}
+
+	latestMerkleRoot, err := s.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
```

Comment:
Thinking out loud about how `DeleteExpired` is going to be called. If it's going to be the same hook that's generating a new deviceEK, it might make sense to have the caller pass in the same root that's being used for the new key, rather than re-fetching it here. I've been kinda sorta trying to minimize the number of calls to "now" / "current root" in EK logic, to avoid the inevitable Heisenbug race conditions that "now" brings with it. (Imagine a bug that only happens if a new root is generated precisely between publication of the new key and deletion of expired ones. There's inevitably going to be things like that, but the fewer different roots we fetch, the less it can happen.)

Reply:
cool i think it's a good change. we might want to change the publishing code to optionally take a root in as well (so you can call device/user gen with the same root and then pass those to `DeleteExpired`

Reply:
Agreed. It might also make testing easier?